## Dataset Goals
1. Determine whether the player is over or undervalued based on statistical rating.
2. Determine if a player is under or over paid based on value and statistical rating.
3. Determine how the individual statistics contributes to the overall statistics of a player.

### Import the data file

In [1]:
import pandas as pd
import numpy as np
df = pd.read_csv("fifa21_raw_data_v2.csv")

C:\Users\admin\AppData\Local\Temp\ipykernel_21420\2199039415.py:3: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("fifa21_raw_data_v2.csv")


### Checking the information of the dataset

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

> From the above, we can observe that there are 77 columns and 18979 rows.

### Determine the size (Number of rows and columns)

In [3]:
df.shape

(18979, 77)

### Browsing through all the columns
> Provides an initial impression and better understanding of the data set.

> In order to make the decision on which data to keep and which to remove.

In [4]:
pd.set_option('display.max_columns', None) # Such that we are able to browse through all columns in the data set.
df.head()

ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  ↓OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33    93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35    92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27    91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29    91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28    91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position        Joined Loan Date End    Value  \
0           Left   93            RW   Jul 1, 2004           NaN  €103.5M   
1          Right   92            ST  Jul 10, 2018           NaN     €63M   
2          Right   91            GK  Jul 16, 2014           NaN    €120M   
3          Right   91           CAM  Aug 30, 2015           NaN    €129M   
4          Right   91            LW   Aug 3, 2017           NaN    €132M   

    Wage Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  €560K        €138.4M        429        85         95                70   
1  €220K         €75.9M        437        84         95                90   
2  €125K        €159.4M         95        13         11                15   
3  €370K          €161M        407        94         82                55   
4  €270K        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioni

*The numerical statistics of the players are not significant data as it does not contribute much to the analysis, hence using overall and potential statistics will suffice.*

### Dropping the statistics columns

In [5]:
updated_df1 = df.drop(df.columns[22:77], inplace=False, axis=1) 
#updated_df1.head()
#updated_df1.shape


In [6]:
updated_df1.drop(updated_df1.columns[np.r_[[0], 2:5, 17:19]], inplace=True, axis=1)

> Removed all individual statistics, ID, LongName, photoUrl, playerUrl, Joined and Loan Date End. 

In [7]:
updated_df1.head()

Name Nationality  Age  ↓OVA  POT                         Club  \
0           L. Messi   Argentina   33    93   93         \n\n\n\nFC Barcelona   
1  Cristiano Ronaldo    Portugal   35    92   92             \n\n\n\nJuventus   
2           J. Oblak    Slovenia   27    91   93      \n\n\n\nAtlético Madrid   
3       K. De Bruyne     Belgium   29    91   91      \n\n\n\nManchester City   
4          Neymar Jr      Brazil   28    91   91  \n\n\n\nParis Saint-Germain   

      Contract   Positions Height Weight Preferred Foot  BOV Best Position  \
0  2004 ~ 2021  RW, ST, CF  170cm   72kg           Left   93            RW   
1  2018 ~ 2022      ST, LW  187cm   83kg          Right   92            ST   
2  2014 ~ 2023          GK  188cm   87kg          Right   91            GK   
3  2015 ~ 2023     CAM, CM  181cm   70kg          Right   91           CAM   
4  2017 ~ 2022     LW, CAM  175cm   68kg          Right   91            LW   

     Value   Wage Release Clause  
0  €103.5M  €560K        €138.4M  
1     €63M  €220K         €75.9M  
2    €120M  €125K        €159.4M  
3    €129M  €370K          €161M  
4    €132M  €270K        €166.5M

> Display the current columns remaining

### Cleaning the format of the Club data

In [8]:
updated_df1["Club"] = updated_df1["Club"].str[4:]

In [9]:
updated_df1.head()

Name Nationality  Age  ↓OVA  POT                 Club  \
0           L. Messi   Argentina   33    93   93         FC Barcelona   
1  Cristiano Ronaldo    Portugal   35    92   92             Juventus   
2           J. Oblak    Slovenia   27    91   93      Atlético Madrid   
3       K. De Bruyne     Belgium   29    91   91      Manchester City   
4          Neymar Jr      Brazil   28    91   91  Paris Saint-Germain   

      Contract   Positions Height Weight Preferred Foot  BOV Best Position  \
0  2004 ~ 2021  RW, ST, CF  170cm   72kg           Left   93            RW   
1  2018 ~ 2022      ST, LW  187cm   83kg          Right   92            ST   
2  2014 ~ 2023          GK  188cm   87kg          Right   91            GK   
3  2015 ~ 2023     CAM, CM  181cm   70kg          Right   91           CAM   
4  2017 ~ 2022     LW, CAM  175cm   68kg          Right   91            LW   

     Value   Wage Release Clause  
0  €103.5M  €560K        €138.4M  
1     €63M  €220K         €75.9M  
2    €120M  €125K        €159.4M  
3    €129M  €370K          €161M  
4    €132M  €270K        €166.5M

In [10]:
updated_df1['Start Year'] = updated_df1['Contract'].str[:4]
updated_df1['End Year'] = updated_df1['Contract'].str[-4:]

In [11]:
updated_df1.head()

Name Nationality  Age  ↓OVA  POT                 Club  \
0           L. Messi   Argentina   33    93   93         FC Barcelona   
1  Cristiano Ronaldo    Portugal   35    92   92             Juventus   
2           J. Oblak    Slovenia   27    91   93      Atlético Madrid   
3       K. De Bruyne     Belgium   29    91   91      Manchester City   
4          Neymar Jr      Brazil   28    91   91  Paris Saint-Germain   

      Contract   Positions Height Weight Preferred Foot  BOV Best Position  \
0  2004 ~ 2021  RW, ST, CF  170cm   72kg           Left   93            RW   
1  2018 ~ 2022      ST, LW  187cm   83kg          Right   92            ST   
2  2014 ~ 2023          GK  188cm   87kg          Right   91            GK   
3  2015 ~ 2023     CAM, CM  181cm   70kg          Right   91           CAM   
4  2017 ~ 2022     LW, CAM  175cm   68kg          Right   91            LW   

     Value   Wage Release Clause Start Year End Year  
0  €103.5M  €560K        €138.4M       2004     2021  
1     €63M  €220K         €75.9M       2018     2022  
2    €120M  €125K        €159.4M       2014     2023  
3    €129M  €370K          €161M       2015     2023  
4    €132M  €270K        €166.5M       2017     2022

### Rearrange the columns

In [12]:
updated_df1 = updated_df1[['Nationality', 'Positions', 'Best Position', 'Name', 'Age', 'Height', 'Weight', 'Club', 'Start Year', 'End Year', '↓OVA', 'POT', 'Value', 'Wage', 'Release Clause' ]]

In [13]:
updated_df1.head()

Nationality   Positions Best Position               Name  Age Height Weight  \
0   Argentina  RW, ST, CF            RW           L. Messi   33  170cm   72kg   
1    Portugal      ST, LW            ST  Cristiano Ronaldo   35  187cm   83kg   
2    Slovenia          GK            GK           J. Oblak   27  188cm   87kg   
3     Belgium     CAM, CM           CAM       K. De Bruyne   29  181cm   70kg   
4      Brazil     LW, CAM            LW          Neymar Jr   28  175cm   68kg   

                  Club Start Year End Year  ↓OVA  POT    Value   Wage  \
0         FC Barcelona       2004     2021    93   93  €103.5M  €560K   
1             Juventus       2018     2022    92   92     €63M  €220K   
2      Atlético Madrid       2014     2023    91   93    €120M  €125K   
3      Manchester City       2015     2023    91   91    €129M  €370K   
4  Paris Saint-Germain       2017     2022    91   91    €132M  €270K   

  Release Clause  
0        €138.4M  
1         €75.9M  
2        €159.4M  
3          €161M  
4        €166.5M

### Converting Value and Wage into Integer Data Type

In [23]:
# To convert the Value Column, keep the current Value column and create a new Value Int column
updated_df1['Value_int'] = updated_df1["Value"].astype("str") # stores a string version of the items in Value
updated_df1['Value_int'] = updated_df1["Value"].str.replace('€','') #.str is to access string functions

updated_df1['Wage_amt'] = updated_df1["Wage"].astype("str")
updated_df1['Wage_amt'] = updated_df1["Wage"].str.replace('€','')

def convert_amount(value):
    if 'M' in value:
        return float(value.replace('M', '')) * 1e6
    elif 'K' in value:
        return float(value.replace('K', '')) * 1e3
    else:
        return float(value)

updated_df1['Value_int'] = updated_df1["Value_int"].apply(convert_amount)
updated_df1['Wage_amt'] = updated_df1["Wage_amt"].apply(convert_amount)

In [24]:
updated_df1.head()

Nationality   Positions Best Position               Name  Age Height Weight  \
0   Argentina  RW, ST, CF            RW           L. Messi   33  170cm   72kg   
1    Portugal      ST, LW            ST  Cristiano Ronaldo   35  187cm   83kg   
2    Slovenia          GK            GK           J. Oblak   27  188cm   87kg   
3     Belgium     CAM, CM           CAM       K. De Bruyne   29  181cm   70kg   
4      Brazil     LW, CAM            LW          Neymar Jr   28  175cm   68kg   

                  Club Start Year End Year  ↓OVA  POT    Value   Wage  \
0         FC Barcelona       2004     2021    93   93  €103.5M  €560K   
1             Juventus       2018     2022    92   92     €63M  €220K   
2      Atlético Madrid       2014     2023    91   93    €120M  €125K   
3      Manchester City       2015     2023    91   91    €129M  €370K   
4  Paris Saint-Germain       2017     2022    91   91    €132M  €270K   

  Release Clause    Value_int  Wage_amt  
0        €138.4M  103500000.0  560000.0  
1         €75.9M   63000000.0  220000.0  
2        €159.4M  120000000.0  125000.0  
3          €161M  129000000.0  370000.0  
4        €166.5M  132000000.0  270000.0

In [25]:
updated_df1 = updated_df1[['Nationality', 'Positions', 'Best Position', 'Name', 'Age', 'Height', 'Weight', 'Club', 'Start Year', 'End Year', '↓OVA', 'POT', 'Value', 'Value_int', 'Wage', 'Wage_amt', 'Release Clause' ]]
updated_df1.head()

Nationality   Positions Best Position               Name  Age Height Weight  \
0   Argentina  RW, ST, CF            RW           L. Messi   33  170cm   72kg   
1    Portugal      ST, LW            ST  Cristiano Ronaldo   35  187cm   83kg   
2    Slovenia          GK            GK           J. Oblak   27  188cm   87kg   
3     Belgium     CAM, CM           CAM       K. De Bruyne   29  181cm   70kg   
4      Brazil     LW, CAM            LW          Neymar Jr   28  175cm   68kg   

                  Club Start Year End Year  ↓OVA  POT    Value    Value_int  \
0         FC Barcelona       2004     2021    93   93  €103.5M  103500000.0   
1             Juventus       2018     2022    92   92     €63M   63000000.0   
2      Atlético Madrid       2014     2023    91   93    €120M  120000000.0   
3      Manchester City       2015     2023    91   91    €129M  129000000.0   
4  Paris Saint-Germain       2017     2022    91   91    €132M  132000000.0   

    Wage  Wage_amt Release Clause  
0  €560K  560000.0        €138.4M  
1  €220K  220000.0         €75.9M  
2  €125K  125000.0        €159.4M  
3  €370K  370000.0          €161M  
4  €270K  270000.0        €166.5M

### Positions in the game

In [ ]:
print(df["Best Position"].unique())
# checking for NA values
print(df["Best Position"].isna().any()) # shows that are no empty values

['RW' 'ST' 'GK' 'CAM' 'LW' 'CB' 'CDM' 'CF' 'CM' 'RB' 'LB' 'LM' 'RM' 'LWB'
 'RWB']
False


### Classify the players by position, ATT, MID, DEF or GK

In [31]:
def position(best_pos):
    if best_pos in ["RW", "LW", "CF", "ST"]:
        return "ATT"
    elif best_pos in ["CAM", "CDM", "CM", "LM", "RM"]:
        return "MID"
    elif best_pos in ["CB", "LB", "RB", "RWB", "LWB"]:
        return "DEF"
    elif best_pos == "GK":
        return "GK"

updated_df1["Field Position"] = updated_df1["Best Position"].apply(position)

In [34]:
updated_df1 = updated_df1[['Nationality', 'Positions', 'Best Position', 'Field Position', 'Name', 'Age', 'Height', 'Weight', 'Club', 'Start Year', 'End Year', '↓OVA', 'POT', 'Value', 'Value_int', 'Wage', 'Wage_amt', 'Release Clause' ]]
updated_df1.head()

Nationality   Positions Best Position Field Position               Name  \
0   Argentina  RW, ST, CF            RW            ATT           L. Messi   
1    Portugal      ST, LW            ST            ATT  Cristiano Ronaldo   
2    Slovenia          GK            GK             GK           J. Oblak   
3     Belgium     CAM, CM           CAM            MID       K. De Bruyne   
4      Brazil     LW, CAM            LW            ATT          Neymar Jr   

   Age Height Weight                 Club Start Year End Year  ↓OVA  POT  \
0   33  170cm   72kg         FC Barcelona       2004     2021    93   93   
1   35  187cm   83kg             Juventus       2018     2022    92   92   
2   27  188cm   87kg      Atlético Madrid       2014     2023    91   93   
3   29  181cm   70kg      Manchester City       2015     2023    91   91   
4   28  175cm   68kg  Paris Saint-Germain       2017     2022    91   91   

     Value    Value_int   Wage  Wage_amt Release Clause  
0  €103.5M  103500000.0  €560K  560000.0        €138.4M  
1     €63M   63000000.0  €220K  220000.0         €75.9M  
2    €120M  120000000.0  €125K  125000.0        €159.4M  
3    €129M  129000000.0  €370K  370000.0          €161M  
4    €132M  132000000.0  €270K  270000.0        €166.5M

In [35]:
updated_df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 18 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Nationality     18979 non-null  object 
 1   Positions       18979 non-null  object 
 2   Best Position   18979 non-null  object 
 3   Field Position  18979 non-null  object 
 4   Name            18979 non-null  object 
 5   Age             18979 non-null  int64  
 6   Height          18979 non-null  object 
 7   Weight          18979 non-null  object 
 8   Club            18979 non-null  object 
 9   Start Year      18979 non-null  object 
 10  End Year        18979 non-null  object 
 11  ↓OVA            18979 non-null  int64  
 12  POT             18979 non-null  int64  
 13  Value           18979 non-null  object 
 14  Value_int       18979 non-null  float64
 15  Wage            18979 non-null  object 
 16  Wage_amt        18979 non-null  float64
 17  Release Clause  18979 non-null 